libs/execution_context/ebpf_maps.c

```diff
@@ -833,34 +928,78 @@ _delete_hash_map(_In_ _Post_invalid_ ebpf_core_map_t* map)
 static void
 _delete_object_hash_map(_In_ _Post_invalid_ ebpf_core_map_t* map)
 {
-    ebpf_core_object_map_t* object_map = EBPF_FROM_FIELD(ebpf_core_object_map_t, core_map, map);
-
     // Release all entry references.
     uint8_t* next_key;
     for (uint8_t* previous_key = NULL;; previous_key = next_key) {
-        uint8_t* value;
+        uint8_t* value = NULL;
         ebpf_result_t result =
             ebpf_hash_table_next_key_pointer_and_value((ebpf_hash_table_t*)map->data, previous_key, &next_key, &value);
         if (result != EBPF_SUCCESS) {
             break;
         }
-        ebpf_id_t id = *(ebpf_id_t*)value;
-        if (id) {
-            ebpf_assert_success(ebpf_object_dereference_by_id(id, EBPF_OBJECT_MAP));
-        }
-    }
 
-    if (object_map->inner_map_template != NULL) {
-        ebpf_object_release_reference(&object_map->inner_map_template->object);
+        __analysis_assume(value != NULL);
+        _clear_map_entry(map, value);
     }
-
     _delete_hash_map(map);
 }
 
+static void
+_object_hash_table_notification(
+    _In_ void* context, _In_ ebpf_hash_table_notification_type_t type, _In_ const uint8_t* key, _In_ uint8_t* value)
+{
+    UNREFERENCED_PARAMETER(key);
+
+    // please refer to the comment in _update_hash_map_entry_with_handle() for this function's usage.
+
+    ebpf_core_object_map_t* map = (ebpf_core_object_map_t*)context;
+    ebpf_id_t object_id = *(ebpf_id_t*)value;
+
+    switch (type) {
+    case EBPF_HASH_TABLE_NOTIFICATION_TYPE_ALLOCATE:
+
+        ebpf_core_object_t* value_object = NULL;
+        ebpf_result_t result = ebpf_object_reference_by_id(object_id, EBPF_OBJECT_MAP, &value_object);
+        if (result != EBPF_SUCCESS) {
+
+            // This can _never_ happen.  We get to this point iff we were already able to grab a ref
+            // (in _update_hash_map_entry_with_handle) so this object _MUST_ exist.  If not, this is a catastrophic
+            // internal error.  Log this as there's nothing we can do by means of recovery.
```

Comment:
if it's a logic bug then suggest asserting (in a debug build)

Reply:
fixed